File: backend/app/routers/opera_pace.py
"""
營運分析 — 訂房 Pace／Pickup：API Router
Prefix: /api/v1/opera/pace

規格：`docs/SPEC_opera_pace.md`

⚠️ 本模組的歷史進度是以**訂房日回推**得出（`opera_reservation_sync` 是整列覆寫、
   無版本），已含後續改期與取消的結果。每個回應的 `source.population` 都會帶
   這句話，畫面必須顯示。

⚠️ 與 `/opera/reservations` 的差別：那邊看的是「訂單**現在**長什麼樣」，
   這邊看的是「某個**過去時點**看到的在手訂房」。多一個 `as_of` 維度。

⚠️ 查詢端點只讀本地資料表，**不打 OHIP**，所以很快也不計費。

⚠️ 全部端點皆為同步 def —— async def 直接呼叫同步 DB 會凍結整站。
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.dependencies import get_current_user, require_permission
from app.models.user import User
from app.services import opera_pace_service as PS

router = APIRouter(dependencies=[Depends(get_current_user)])
_VIEW = require_permission("opera_pace_view")


def _bad(e: ValueError) -> HTTPException:
    return HTTPException(status_code=400, detail=str(e))


@router.get("/data-range", summary="資料涵蓋範圍與快照就緒度")
def data_range(db: Session = Depends(get_db), _: User = Depends(_VIEW)):
    """涵蓋範圍、無法定位時點的取消筆數、快照累積天數。

    ⚠️ 入住日區間是**未來導向**，前端請用 antd 原生 RangePicker，
       不要用 StandardRangePicker（CLAUDE.md §8.4）。
    """
    return PS.data_range(db)


@router.get("/readiness", summary="快照累積天數（Phase 2 就緒度）")
def readiness(db: Session = Depends(get_db), _: User = Depends(_VIEW)):
    """快照精確版還需要累積幾天。只讀本地表，不打 OHIP。"""
    return PS.readiness(db)


@router.get("/curve", summary="訂房曲線（含去年同期）")
def curve(
    stay_date: str = Query(..., description="ISO YYYY-MM-DD"),
    compare: str = Query("weekday", pattern="^(weekday|date)$"),
    max_lead: int = Query(120, ge=7, le=365),
    db: Session = Depends(get_db), _: User = Depends(_VIEW),
):
    """單一入住日的 OTB 隨觀察日演進，附去年同期對照曲線。

    🎯 去年同期曲線用同一套回推邏輯即可得出 —— 快照路線要等滿一年。
    """
    try:
        return PS.curve(db, stay_date=stay_date, compare=compare, max_lead=max_lead)
    except ValueError as e:
        raise _bad(e)


@router.get("/otb-matrix", summary="入住日 × 提前期 OTB 矩陣")
def otb_matrix(
    start: str = Query(..., description="入住日起（ISO）"),
    end: str = Query(...),
    leads: str = Query("90,60,30,14,7,3,1,0", description="逗號分隔的提前期"),
    compare: str = Query("weekday", pattern="^(weekday|date)$"),
    as_of: str | None = Query(None, description="觀察日，預設今天"),
    window: int = Query(7, description="pickup 欄的觀察窗：1／3／7／14"),
    db: Session = Depends(get_db), _: User = Depends(_VIEW),
):
    """主表格。⚠️ 觀察日還沒到的格子回 `null` 而不是 0。

    ⚠️ `as_of` 未帶時預設今天。看歷史區間時務必帶上，否則 pickup 欄永遠是 0
       （「最近 7 天」對兩年前的入住日不會有任何異動）。
    """
    try:
        ls = tuple(sorted({int(x) for x in leads.split(",") if x.strip()},
                          reverse=True))
        if not ls or any(x < 0 or x > 365 for x in ls):
            raise ValueError("leads 需為 0～365 的整數")
        return PS.otb_matrix(db, start=start, end=end, leads=ls, compare=compare,
                             as_of=as_of, window=window)
    except ValueError as e:
        raise _bad(e)


@router.get("/pickup", summary="逐入住日 Pickup（新增／取消／淨）")
def pickup(
    start: str = Query(...), end: str = Query(...),
    window: int = Query(7, description="觀察窗天數：1／3／7／14"),
    as_of: str | None = Query(None, description="觀察日，預設今天"),
    db: Session = Depends(get_db), _: User = Depends(_VIEW),
):
    """🎯 新增與取消**分開列** —— 淨值相同不代表狀況相同。

    回應含 `verify`：OTB 差值應等於淨 pickup，不符會列在 warnings。
    """
    try:
        return PS.pickup(db, start=start, end=end, window=window, as_of=as_of)
    except ValueError as e:
        raise _bad(e)


@router.get("/pickup/dimension", summary="維度別 Pickup（參考值）")
def pickup_dimension(
    start: str = Query(...), end: str = Query(...),
    dimension: str = Query("market_code"),
    window: int = Query(7), as_of: str | None = Query(None),
    db: Session = Depends(get_db), _: User = Depends(_VIEW),
):
    """⚠️ 維度取自訂房**目前**的房型／通路，中途改過會算在新維度上。
    回應帶 `is_reference_only=true`，畫面必須標「參考值」。
    """
    try:
        return PS.pickup_dimension(db, start=start, end=end,
                                   dimension=dimension, window=window, as_of=as_of)
    except ValueError as e:
        raise _bad(e)


@router.get("/day-detail", summary="單日 Pickup 組成明細（Drawer）")
def day_detail(
    stay_date: str = Query(...), window: int = Query(7),
    as_of: str | None = Query(None),
    db: Session = Depends(get_db), _: User = Depends(_VIEW),
):
    """觀察窗內新增／取消了哪些訂單。

    ⚠️ 非 Ragic 來源，`ragic_url` 一律為空字串（CLAUDE.md §7 該欄不適用）。
    """
    try:
        return PS.day_detail(db, stay_date=stay_date, window=window, as_of=as_of)
    except ValueError as e:
        raise _bad(e)

File: frontend/src/api/operaPace.ts
/**
 * 營運分析 — 訂房 Pace／Pickup API 封裝
 * Prefix: /api/v1/opera/pace
 *
 * ⚠️ **這一頁的歷史進度是「回推」出來的**：訂房同步是整列覆寫、無版本，
 *    所以我們只有「每筆訂房現在長什麼樣」。本模組用 booking_date /
 *    cancellation_date 把現況往回切，已含後續改期與取消的結果。
 *    `source.population` 會把這句話帶到畫面上 —— 不要拿掉。
 *
 * ⚠️ 與 `api/operaReservation.ts` 的差別：那邊看訂單「現在」長什麼樣，
 *    這邊多一個 `as_of`（觀察時點）維度。
 *
 * ⚠️ 只讀本地資料表，不打 OHIP，不計費。
 */
import apiClient from '@/api/client'

const PACE = '/opera/pace'

export type PaceDimension =
  | 'market_code' | 'room_type' | 'channel' | 'rate_code' | 'source_code'

export type CompareMode = 'weekday' | 'date'

export interface Coverage {
  filled: number
  total: number
  ratio: number | null
  /** 填充率 < 50%，畫面必須加警語 */
  is_low: boolean
}

export interface PaceSource {
  provider: string
  hotel_id: string
  /** ⚠️ 回推失真說明，必須顯示在畫面上 */
  population: string
  precision: string
  note: string
}

export interface SnapshotReadiness {
  distinct_snapshot_days: number
  first_snapshot_date: string | null
  last_snapshot_date: string | null
  business_days_with_final: number
  required_days: number
  ready: boolean
  remaining_days: number
}

export interface PaceDataRange {
  start: string | null
  end: string | null
  last_past: string | null
  earliest_booking_date: string | null
  has_data: boolean
  /** 狀態是取消但無取消日期，無法定位時點，已從所有觀察點排除 */
  unresolved_cancels: number
  snapshot: SnapshotReadiness
  source: PaceSource
}

export interface CurvePoint {
  lead_days: number
  as_of: string
  room_nights: number
  room_revenue: number | null
  adr: number | null
  ly_room_nights: number | null
  vs_ly: number | null
  vs_ly_rooms: number | null
}

export interface CurveResult {
  stay_date: string
  compare: CompareMode
  ly_stay_date: string
  points: CurvePoint[]
  final: number
  ly_final: number
  unresolved_cancels: number
  /** ⚠️ 沒有訂房日的房晚數，無法回推故排除。這是本頁與訂房分析對不起來的原因，必須顯示 */
  missing_booking_date: number
  source: PaceSource
}

export interface OtbRow {
  stay_date: string
  weekday: number
  is_weekend: boolean
  lead_days_now: number
  /** key 是 lead 天數字串；⚠️ null = 觀察日還沒到，不是 0 */
  otb: Record<string, number | null>
  room_nights: number
  room_revenue: number | null
  adr: number | null
  /** 以 as_of 為右端點、往前 window 天的淨 pickup */
  pickup_net: number
  pickup_new: number
  pickup_cancels: number
  ly_stay_date: string
  ly_room_nights: number
  vs_ly: number | null
}

export interface OtbMatrixResult {
  range: { start: string; end: string }
  leads: number[]
  compare: CompareMode
  /** 實際採用的觀察日（未帶參數時是今天） */
  as_of: string
  window: number
  rows: OtbRow[]
  summary: { room_nights: number; ly_room_nights: number; pickup_net: number }
  unresolved_cancels: number
  /** ⚠️ 沒有訂房日的房晚數，無法回推故排除。這是本頁與訂房分析對不起來的原因，必須顯示 */
  missing_booking_date: number
  revenue_coverage: Coverage
  source: PaceSource
}

export interface PickupRow {
  stay_date: string
  weekday: number
  gross_new: number
  cancels: number
  net: number
  otb_before: number
  otb_after: number
  /** 恆等式自我檢查：OTB 差值 == 淨 pickup */
  verified: boolean
}

export interface PickupResult {
  range: { start: string; end: string }
  window: number
  from: string
  to: string
  rows: PickupRow[]
  summary: { gross_new: number; cancels: number; net: number }
  unresolved_cancels: number
  /** ⚠️ 沒有訂房日的房晚數，無法回推故排除 */
  missing_booking_date: number
  verify: { all_passed: boolean; warnings: string[] }
  source: PaceSource
}

export interface PickupDimensionResult {
  range: { start: string; end: string }
  dimension: PaceDimension
  window: number
  from: string
  to: string
  rows: Array<{
    key: string; gross_new: number; cancels: number
    net: number; otb_after: number
  }>
  unresolved_cancels: number
  /** ⚠️ 沒有訂房日的房晚數，無法回推故排除 */
  missing_booking_date: number
  coverage: Coverage
  /** ⚠️ 恆為 true：維度取自訂房目前狀態，畫面必須標「參考值」 */
  is_reference_only: boolean
  source: PaceSource
}

export interface DayDetailItem {
  confirmation_no: string
  booking_date: string
  cancellation_date: string
  cancellation_reason_code: string
  arrival: string
  departure: string
  nights: number | null
  lead_days: number | null
  resv_status: string
  market_code: string
  room_type: string
  rate_code: string
  channel: string
  room_revenue: number | null
  /** 非 Ragic 來源，恆為空字串（CLAUDE.md §7 該欄不適用） */
  ragic_url: string
}

export interface DayDetailResult {
  stay_date: string
  window: number
  from: string
  to: string
  added: DayDetailItem[]
  cancelled: DayDetailItem[]
  summary: { gross_new: number; cancels: number; net: number }
  source: PaceSource
}

const get = async <T>(path: string, params?: any): Promise<T> =>
  (await apiClient.get(`${PACE}${path}`, { params })).data

export const fetchPaceDataRange = () => get<PaceDataRange>('/data-range')

export const fetchPaceReadiness = () => get<SnapshotReadiness>('/readiness')

export const fetchPaceCurve = (p: {
  stay_date: string; compare?: CompareMode; max_lead?: number
}) => get<CurveResult>('/curve', p)

/** ⚠️ 看歷史區間務必帶 as_of，否則 pickup 欄永遠是 0 */
export const fetchOtbMatrix = (p: {
  start: string; end: string; leads?: string; compare?: CompareMode
  as_of?: string; window?: number
}) => get<OtbMatrixResult>('/otb-matrix', p)

export const fetchPacePickup = (p: {
  start: string; end: string; window?: number; as_of?: string
}) => get<PickupResult>('/pickup', p)

export const fetchPacePickupDimension = (p: {
  start: string; end: string; dimension: PaceDimension
  window?: number; as_of?: string
}) => get<PickupDimensionResult>('/pickup/dimension', p)

export const fetchPaceDayDetail = (p: {
  stay_date: string; window?: number; as_of?: string
}) => get<DayDetailResult>('/day-detail', p)
